Make cast_to a static member of Object.

hpvb opened the pull request:
Currently we rely on some undefined behavior when Object->cast_to() gets
called with a Null pointer. This used to work fine with GCC < 6 but 
newer versions of GCC remove all codepaths in which the this pointer is
Null. However, the non-static cast_to() was supposed to be null safe.

This patch makes cast_to() Null safe and removes the now redundant Null
checks where they existed.

The following demos have been imported and run without issues on Fedora 26/GCC 7.1.1:

* 2d/lights_and_shadows
* 2d/physics_platformer
* 2d/platformer
* 3d/kinematic_character
* 3d/material_testers
* 3d/platformer

This validates that it at least addresses #10517 and #10515 also. These were already fixed by this commit removes those null checks again and the crashes are still gone. This suggests to me that this will likely fix other random crashes on GCC6+ also.

Zireael07 commented:
Man I think you messed up git somehow, what are all those commits doing there?

hpvb commented:
@Zireael07 yeah, I accidentally rebased on my fork rather than master, sorry :) fixed now

RandomShaper commented:
I've not checked the diff thoroughly, but your idea is nice.

But then I wonder if commits adding null checks just because of the UB issue should be reverted, as no longer meaningful.

akien-mga commented:
This commit already removes the redundant null checks, e.g. https://github.com/godotengine/godot/pull/10581/commits/0f4d1adb62cb4cb96aac63fe67f8f4c58b8dae7d#diff-6435cbec06d56101c9bcc0061478fae7L276

RandomShaper commented:
Terrific. :)

Zylann commented:
Just a note, I thought dereferencing a null pointer was fine as long as `this` isn't accessed? I last saw this trick in the Qt codebase

hpvb commented:
@Zylann I think we're hitting the problem explained in this blogpost:  https://www.viva64.com/en/b/0226/

hpvb commented:
@Zylann This post copy/pastes some part of the relevant standard: https://www.viva64.com/en/w/V704/

hpvb committed:
Add a static version of Object::cast_to()

This is to prepare to replace all instances of the member version of
cast_to().

hpvb committed:
Convert Object::cast_to() to the static version

Currently we rely on some undefined behavior when Object->cast_to() gets
called with a Null pointer. This used to work fine with GCC < 6 but
newer versions of GCC remove all codepaths in which the this pointer is
Null. However, the non-static cast_to() was supposed to be null safe.

This patch makes cast_to() Null safe and removes the now redundant Null
checks where they existed.

It is explained in this article: https://www.viva64.com/en/b/0226/

akien-mga commented:
I can't wait not to notice the 50 potential bugs this will have magically fixed :P
